### PR TITLE
CRM-18792 - Create CSS theming subsystem

### PR DESCRIPTION
**Pull request by totten**
- [CRM-18792: Allow extensions to define CSS themes](https://issues.civicrm.org/jira/browse/CRM-18792)

### CODE REVIEW

**Comment by totten** [comment]
So in this case, the test failure really is indicating a change in behavior. Consider the call:

```php
Civi::resources()->addStyleFile('civicrm', 'css/bootstrap/does-not-exist.css');
```

The file `css/bootstrap/does-not-exist.css` does not exist. This is a sketchy situation, and the PR changed the behavior:
 * The old behavior (expected by `CRM_Core_ResourcesTest`) is to add the `<style>` regardless.
 * The new behavior (expected by `Civi\Core\ThemeTest`) is to search several places (subtheme; parent theme; core folder); if no one has a copy of `css/bootstrap/does-not-exist.css`, then skip it.

It's not really clear what behavior makes the most sense here (throw an exception; add the file anyway; skip it). In absence of a clear reason to change, I'll update PR to match the old behavior.

**Comment by totten** [comment]
@AkA84 @agh1 @jamienovick @kcristiano , I've drafted a chapter for https://docs.civicrm.org to go along with this:

https://github.com/totten/civicrm-dev-docs/blob/master-theme/docs/theme.md
https://github.com/civicrm/civicrm-dev-docs/pull/8

The docs have a some big blank spaces for describing the actual CSS of `civicrm.css` and `bootstrap.css`. Whenever we get URL for a style-guide for each scheme

**Comment by jamienovick** [comment]
Ill perhaps leave this to the more technical of you to review but thanks 
for doing this!

Once it all looks good and you guys are happy I'll draft a blog post for 
the more non techies...

J

On behalf of Compucorp Ltd

Jamie Novick / Director

Tel: +44 (0)207 096 3336 / Mob: +44 (0)775 123 8875

Skpe ID: jamie.novick
www.compucorp.co.uk<http://www.compucorp.co.uk/>

------------------------------------------------------------------------

This email is confidential and is intended for the addressee only. If 
you are not the addressee, please delete the email and do not use it in 
any way. Compucorp Ltd does not accept or assume responsibility for any 
use of or reliance on this email by anyone, other than the intended 
addressee to the extent agreed in the relevant contract for the matter 
to which this email relates (if any).

Unless expressly stated to the contrary, the content of this e-mail does 
not represent the professional advice, views or opinion of Compucorp 
Ltd. Compucorp Ltd. reserve the right to monitor all business and 
personal e-mails to and from its associated e-mail addresses. By 
responding to this e-mail you confirm that you consent to this action.

All reasonable efforts have been made to confirm that this e-mail and 
any attachments are virus free but you are advised to check. Compucorp 
Ltd. does not accept any responsibility for any damage caused by any 
transmission to the recipient's computer systems.

------------------------------------------------------------------------

_

**Comment by AkA84** [comment]
Hi @totten , I’ve read the documentation and I’ve got a couple of questions

1.
Are you going to add a theme-specific class to the main container tag? Something like: 

```html
 <div class="crm-container crm-theme-shoreditch"> </div> 
```
```html
 <div class="crm-container crm-theme-greenwich"> </div> 
```
as discussed during the sprint I think this would provide a nice way to write ad-hoc css rules just for a specific theme
 
_(btw: I just realized that `div.crm-container` - at least in the admin - is way deep into the page, past the navbar and the title bar. As such we should probably apply the theme class to the `<body>` directly)_

2.
I thought we were going to have a subfolder for the theme even if it’s the only theme in the extension?   

For example, from the "Packaging section": 
```
 org.civicrm.theme.newyork/info.xml
 org.civicrm.theme.newyork/newyork.php
 org.civicrm.theme.newyork/newyork/css/civicrm.css 
org.civicrm.theme.newyork/newyork/css/bootstrap.css 
```
this way a theme author would have to learn just a single, coherent folder structure for both single and multi theme extensions, and I think at that point you could get rid of the `prefix` property in the hook, making its "interface" simpler 
```php
// FILE: newyork.php
function newyork_civicrm_themes(&$themes) {
  $themes['astoria'] = array(
    'ext' => 'org.civicrm.theme.newyork',
    'title' => 'Astoria'
  );
  $themes['wallstreet'] = array(
    'ext' => 'org.civicrm.theme.newyork',
    'title' => 'Wall Street',
  );
}
```
```
org.civicrm.theme.newyork/info.xml
org.civicrm.theme.newyork/newyork.php
org.civicrm.theme.newyork/astoria/css/civicrm.css
org.civicrm.theme.newyork/astoria/css/bootstrap.css
org.civicrm.theme.newyork/wallstreet/css/civicrm.css
org.civicrm.theme.newyork/wallstreet/css/bootstrap.css
```
---

```php
// FILE: newyork.php
function newyork_civicrm_themes(&$themes) {
  $themes['astoria'] = array(
    'ext' => 'org.civicrm.theme.newyork',
    'title' => 'Astoria',
    'search_order' => array('astoria', '_newyork_common_', '_fallback_'),
  );
  $themes['wallstreet'] = array(
    'ext' => 'org.civicrm.theme.newyork',
    'title' => 'Wall Street',
    'search_order' => array('wallstreet', '_newyork_common_', '_fallback_'),
  );
  $themes['_newyork_common_'] = array(
    'ext' => 'org.civicrm.theme.newyork',
    'title' => 'New York (Base Theme)',
  );
  // Note: "_newyork_common_" begins with "_".  It is a hidden, abstract
  // theme which cannot be directly activated.
}
```
```
org.civicrm.theme.newyork/info.xml
org.civicrm.theme.newyork/newyork.php
org.civicrm.theme.newyork/newyork-common/css/civicrm.css
org.civicrm.theme.newyork/astoria/css/bootstrap.css
org.civicrm.theme.newyork/wallstreet/css/bootstrap.css
```

What do you think?

3.
In the "Packaging" section we have
> Exclude Files: If the the CMS theme already loads a copy of bootstrap.css through the CMS, then it may be redundant to load a copy of bootstrap.css through Civi's theming layer.

In the "Advanced: `excludes`" section we have
> For example, if you have provided styling rules through a CMS theme, then loading civicrm.css could be redundant.

Two questions:
1) Do I understand correctly that the scenarios described above seems to apply more for a theme "consumer" (as in, a developer managing his own CiviCRM instance) rather than a theme author? Because the latter can’t possibly know if a CiviCRM instance that is using his theme is already, say, using its own `bootstrap.css`, and as such needs to exclude the one in the extension theme.

Basically I’m not sure I understand what’s the use case of `excludes` for a theme author (if there was meant to be any!), while I can imagine someone managing his own CiviCRM instance who amends the theme extension hook once installed to exclude files that he, in his particular case, doesn’t need

2) Could you explain to me the `civicrm:css/bootstrap.css` syntax?

4.
Are we going to have a this stage a parent/child theme mechanism?

Thanks!
Alessandro

**Comment by totten** [comment]
(Note: I've tried to re-order questions to get the simpler ones out first.)

> Could you explain to me the civicrm:css/bootstrap.css syntax?

Well, if you compare to the call `addStyleFile('civicrm','css/bootstrap.css')`, then we're just joining those two parts by `:`.  But on second thought, it would make more sense for 'excludes' to match the file-naming convention.  (e.g.  `css/bootstrap.css` for the typical case; `org.civicrm.volunteer-style/nonstandard.css` for the oddball case). So I can patch that.

> Are we going to have a this stage a parent/child theme mechanism?

That's what `search_order` was intended for. But maybe you had something else in mind?

>  Are you going to add a theme-specific class to the main container tag?
> ... we should probably apply the theme class to the <body> directly

Thanks for catching/remembering! I'll try to do something for the container tag. A few thoughts:

 * Applying another class to the `crm-container` should be pretty straightforward, but the `<body>` is technically trickier.  (Each CMS has a different way of composing the page.) 
 * We could add a `<script>` in the `<head>` to say something like `document.body.className += ' crm-theme-greenwich';`. That's simple and already supported by our UF abstraction, but it *might* be time-sensitive (e.g. flashing when the JS runs).
 * Could you describe a bit more about the use-case for `<body class="crm-theme-greenwich">`? (Bear in mind that some screens involve Civi... and some screens are purely driven by CMS. They don't load Civi at all.)
 * We could add the marker to both `crm-container` and `body`. 

> I thought we were going to have a subfolder for the theme even if it’s the only theme in the
> extension?

TBH, I feel like we got into bike-shed territory on this, but I think I understand the design you described, so I've tried to make a gist of the different approaches (with some supportive/opposing arguments; feel free to fork/edit if I've misinterpreted or misrepresented).

https://gist.github.com/totten/cf80ebf266185cb0b9fbd7f463ad2515

Personally, I'm suspicious about getting the semantics right (how parent/child should work; how to override CSS vs how to supplement CSS; how to convey the semantics to downstream devs; etc). I'd like to get *single* theme working before getting really invested in multitheme, so  my gut here says "keep it simple/less is more/let downstream iterate on it". If I could rate each approach (0=awful, 5=wonderful):

 * "1-optional-prefix.php" -- Acceptable, 3.0/5.0
 * "2-fixed-structure.php" -- OKish, 2.5/5.0
 * "3-just-the-callback.php" -- Fairly good, 4.0/5.0

How would you rate them, and how much do you think it matters? Perhaps a straw-poll (pinging a few more devs) would be useful?

> In the "Advanced: excludes" section ...  Do I understand correctly that
> the scenarios described above seems to apply more for a theme "consumer"
> (as in, a developer managing his own CiviCRM instance) rather than a theme
> author?  Because the latter can’t possibly know if a CiviCRM instance that
> is using his theme is already, say, using its own bootstrap.css, and as
> such needs to exclude the one in the extension theme.

The scenario I had in mind was this: Alice is a theme developer who's written a Drupal theme based on Bootstrap and SASS. She wants the theme to work nicely in Civi, and it would mostly work (because it defines all the standard Bootstrap classes), and she's willing to add a few extra CSS classes or SASS directives to satisfy Civi's style-guide. All her code compiles into one `alice-bootstrap.css` file, and the CMS system loads that file on every page-request. Civi never needs to load `bootstrap.css`, so she `excludes` it.

Of course, this is one scenario, and I think you're describing another -- Bob is a Drupal themer who publishes a Bootstrap-based theme. Carol is an admin who installs both Bob's theme *and* installs CiviCRM. In this case, Bob's theme is close (it already loads `bootstrap.css` on all pages), but it doesn't fully comply with the style-guide (missing a few extra Civi-specific classes). In this scenario, we need to either use a polyfill/shim (to define the missing CSS classes) or fully overload the CSS (within the limited scope of Civi's content area).

IMHO, the current PR *allows* a themer to address any of those scenarios, e.g. 

 * Make an all-encompassing theme which addresses requirements of Bootstrap, CMS, and Civi. (Package this as a CMS theme. The CMS loads a copy of `bootstrap.css`, so it should disable Civi's reference to `bootstrap.css` by using `excludes`.)
 * Make a full overload theme. (Package this as a Civi extension. The file `css/bootstrap.css` is compiled from Bootstrap's SASS and also includes some Civi-specific classes.)
 * Make a polyfill theme. (Package this as a Civi extension. The file `css/bootstrap.css` just defines new, Civi-specific classes... because the CMS has already loaded vanilla Bootstrap.)

I like that this allows each scenario, but it does put a higher burden on the administrator -- when configuring `theme_backend` and `theme_frontend`, they need to choose CMS and Civi themes based on compatibility. It's easy to mix options and screw up the UI.

For reference... well thinking about this... I drew a couple diagrams which might be interesting. The first is a Venn diagram of the different CSS vocabularies:

https://docs.google.com/drawings/d/1UeTddVvmHBCIEwQ-NoNTWoNWgvs1Nvq1kdwe2tif1xA/edit?usp=sharing

and the second shows the ways to package up themes:

https://docs.google.com/drawings/d/15wfwOCJ6YsThnyP7TATC1CZt9YCyyd-o1RuzfD57ZSI/edit?usp=sharing

Not sure yet how to make the admin experience smoother, so will try to sleep on that.

**Comment by AkA84** [comment]
About parent themes:

_(bear in mind that this is merely a brainstorming thing! Might be a bit too much for the first iteration of the theming system, but thought might be useful to at least bring up this possible feature)_

I think `search_order` is fine for handling shared assets between different variations of the same theme (or of different themes in the same extension, if one wants so)

What we were briefly discussing during the sprint was to have a system where a theme extension can reference another theme extension as its own parent.

So let’s say you have `org.civicrm.theme.europe` and `org.civicrm.theme.italy`, where the latter uses the former as its foundation:
```php
// In org.civicrm.theme.europe/europe.php
$theme["europe"] = array(
  "ext" => "org.civicrm.theme.europe",
  "title" => "Europe",
);

// In org.civicrm.theme.italy/italy.php
$theme["italy"] = array(
  "ext" => "org.civicrm.theme.italy",
  "title" => "Italy",
  "parent" => "europe"
);
```
If the current theme is set to be `europe`, then the following resources will be loaded:
```
org.civicrm.theme.europe/europe/civicrm.css
org.civicrm.theme.europe/europe/bootstrap.css
```
If the current theme is set to be `italy`, then the following resources will be loaded:
```
org.civicrm.theme.europe/europe/civicrm.css
org.civicrm.theme.europe/europe/bootstrap.css
org.civicrm.theme.italy/italy/civicrm.css
org.civicrm.theme.italy/italy/bootstrap.css
```
(if then CiviCRM could concatenate the files it would be even better)

Two notes about this solution:
1. Of course the dependency between themes can be achieved manually anyway, by just importing via SASS the parent theme (like we’re doing for the [CiviHR theme](https://github.com/civicrm/civihr/blob/develop/org.civicrm.bootstrapcivihr/scss/style.scss#L1-L2)), but this way you would have CiviCRM "officially" supporting parent/child relationship themes
2. I think this idea originally came up by mentioning the Drupal’s [sub-theming system](https://www.drupal.org/node/225125), but notice that the main difference (at least in this first draft) would be that the civicrm system would just add the child theme resources to the parent’s one, while drupal’s instead make the child theme overrides the parent 

Come to think of it, I wonder if the `parent` property could be anyway a better substitute for `search_order` (not using the `prefix` property as per my comment #2) in terms of simplicity of use

```php
function newyork_civicrm_themes(&$themes) {
  $themes["_newyork_common_"] = array(
    "ext" => "org.civicrm.theme.newyork",
    "title" => "New York (Base Theme)",
  );
  $themes["astoria"] = array(
    "ext" => "org.civicrm.theme.newyork",
    "title" => "Astoria",
    "parent" => "_newyork_common_"
  );
  $themes["wallstreet"] = array(
    'ext' => "org.civicrm.theme.newyork",
    "title" => "Wall Street",
    "parent" => "_newyork_common_"
  );
}
```

**Comment by totten** [comment]
@AkA84 thanks for the comments. Will try to consider them more.

Question for clarification -- in describing the relationship between https://github.com/civicrm/civihr/blob/develop/org.civicrm.bootstrapcivihr/ vs https://github.com/civicrm/civihr/blob/develop/org.civicrm.bootstrapcivicrm/ , I'm a little confused about whether it's doing:

 1. *Extending the CSS vocabulary* -- `bootstrapcivihr` adds new items to the style-guide/pattern-library beyond the items already supported by `bootstrapcivicrm`. (This was my impression from the Colorado discussions.)
 2. *Altering the look* -- `bootstrapcivihr` changes the colors, padding, or other visual effects of elements already in `bootstrapcivicrm`. (My impression is that subtheming is usually intended for this.)

**Comment by AkA84** [comment]
@totten I didn't expect you to reply so quickly :D I'm still in the middle of adding my replies to the remaining comments, bear with me :)

**Comment by AkA84** [comment]
Hi @totten , I finished to write my replies in the comment above ^

**Comment by jamienovick** [comment]
re: CMS+Civi themes approach

Im not 100% sure what the debate is about but I think we should have a "CIVICRM admin Drupal theme" that is neutral and simply sets out that standard regions. Then all the clever stuff is done in the CiviCRM theme layer itself.

nb. Is the public/private CiviCRM pages being considered? I think it is important that we avoid putting out default styles for public pages (or are easily able to switch these off!).

**Comment by adixon** [comment]
Thanks for all this. Trying to make Civi play nicely with an unknown CMS + theme is tricky ...

I don't have many strong opinions about how it's done (though I'll probably complain at the end of it all anyway), but here is what I see as the key challenge: 2 conflicting and not-well-defined needs.

Need 1. Out of the box, the Civi 'admin' experience looks reasonable and functional for both desktop and mobile.

So for those screens, civi can be opinionated and aggressive, and anyone who wants to change it can expect to do some work. Civi will need to provide all the bootstrap bits and assume relatively little from the CMS.

Need 2. It's possible and not to painful to make the civi 'public' experience relatively 'seamless' with the rest of the site.

I don't expect plug and play here, but the current situation where the default can be hugely jarring and it's often a royal pain to get it even reasonably non-jarring is a big pain point, especially for (e.g.) new themers coming from Drupal.

For this to work, Civi has to be very unopinionated and flexible, and it would seem to me that in this case we don't want to be providing all the bootstrap bits, but only generating html with bootstrappish classes that look nice in a CMS bootstrap theme but isn't impossibly nasty in another theme.

In particular - what about when the CMS is using a slightly different version of bootstrap compared with the Civi version? I know we can just do some clever bits so they co-exist, but that's not very nice integration and makes theming extra complicated.

...

The distinction between 'public' and 'admin' is one vague area, though the civitheme in Drupal seems to have a reasonable working definition. I'm not sure if/how Wordpress or Joomla handle this.

I've also recently waded through a similar discussion for drupal 8, which is nicely summarized here: https://www.lullabot.com/articles/a-tale-of-two-base-themes-in-drupal-8-core
Basically, they had two schools of thought about what the CMS should provide and ended up with two core themes to handle both schools in a clever and coherent way.

This problem is slightly different, but some kind of clarity about the two separate needs and how they are resolved with this solution would ease my mind.

**Comment by AkA84** [comment]
Coming back to this after so long certainly helped me with approaching it with a fresh mind!

Biggest problem for me was that I couldn't understand the `exclude` functionality in the context of a [Drupal-packaged](https://github.com/totten/civicrm-dev-docs/blob/master-theme/docs/theme.md#packaging-drupaljoomlawordpress) theme, given that I thought that hooks could be used only in Civi extension, so the Alice's scenario was a bit confusing. Re-reading the guide now I see that you can use the hooks also in a Drupal/Joomla/Wordpress module (although it seems this hadn't been properly tested yet?), so I understand how you would package your theme with an additional Civi-specific bit of configuration so that you can tell Civi not to try to load a `bootstrap.css` file!

So ok, this solution takes into account all three main scenarios, which is good, but I was wondering about the pros and cons of each of them. And also we should agree and be clear on what approach the initial theme ("shoreditch") should follow.

(note: this considers for now only the admin part, i'll touch on the 'public' one later)

### All-encompassing approach
**Pros:** Given that you've disabled the Civi admin css and you're relying entirely on the CMS admin theme, you don't have to worry about the two conflicting with each other.

**Cons:** How common would this be? Would it be common for themers to package their Drupal/Wordpress/Joomla themes with extra bits and pieces specific for Civi? Would these themes still be distributed as generic CMS themes or would they be more niche, just for the Civi community? Also this seems a lot of work for the initial theme, since we would have to create 3 versions of it for the 3 CMS, not to mention that in this case it would be a full CMS admin theme, not just a Civi theme

### Polyfill approach
**Pros:** It's a great way to leverage already existing Bootstrap themes out there, so that you don't need to re-invent the wheel for the common Bootstrap elements and you can focus on the CiviCRM custom elements only. I imagine one would package a polyfill theme extension with _info.xml_ explicitly saying that the extension requires a specific CMS theme to be enabled, and that would be it. Given that the theme has a very specific CMS theme as a dependency, you can either avoid the conflicts between the two altogether or know exactly what they are and fix them

**Cons:** You are introducing a big external dependency on your Civi theme, and as @adixon says, probably Civi, at least for the admin part, should assume very little from the CMS and be very opinionated. For the initial theme this would mean either finding a good CMS admin theme candidate or, again, creating one our own

### Overload approach
This is the approach that we are taking now with the theme we are developing here at Compucorp and also the approach that was mostly discussed in previous conversations (see my previous comment)

**Pros:** With this approach you author the Civi theme without having to make any assumptions on what CMS theme the admin has enabled on the site, which means you have no dependency.

**Cons:** Since the Civi theme is completely independent, it will have to reset any interfering style coming from the CMS, and given that you can't know what theme is enabled, it would be tricky to say the least to be sure that your theme will work fine in every scenario (Unless you introduce a dependency in the form of a "blank" CSM admin theme?)

Does this sound correct in relation with what has been discussed so far? Or did I miss/misunderstand something?

About the public part of the site, I'm afraid I'm not yet very familiar with how Civi works in that regards? Does it inject its css in it (if i take a look at the [demo](http://dmaster.demo.civicrm.org/) I don't see any sign of that)? What are your biggest concerns about it?


Hope this was useful somehow, at least to keep the conversation going!

**Comment by eileenmcnaughton** [comment]
@totten this is now officially the oldest PR & is thoroughly stale. Is anything happening with it. I do think that closing the PR & just linking to it from the JIRA is a good way to deal with good intentions / good ideas that didn't quite get somewhere

**Comment by jamienovick** [comment]
Does the new shoreditch theme have this as a dependency @totten @AkA84 ?

**Comment by eileenmcnaughton** [comment]
@totten this is now the oldest ticket!

What part of this needs merging? I know shoreditch is already released...

**Comment by colemanw** [comment]
This is a year old and Shoreditch is progressing without it. Let's close.

**Comment by jusfreeman** [comment]
Just linking to what I think is the related feature request, https://lab.civicrm.org/dev/core/issues/378 - where discussion about this change can continue.
